Allow entities to stop by terminating without remember entities

johanandren opened the pull request:
References #29383

We missed an allowed transition from running/active to stopped/NoState in shard
when the logic was rewritten.

johanandren committed:
Allow entities to stop by terminating in sharding without remember entities #29383

We missed an allowed transition from running/active to stopped/NoState in shard
when the logic was rewritten.

raboof reviewed (approved):
Ouch... surprising no existing tests picked up on this.

johanandren commented:
There was a test exercising it but since sharding kind of self healed and there was no specific post condition to check if the shard crashed it was not noticed.

I'm adding a toggle for crash vs just log error on top of this in a bit.

johanandren committed:
Add a toggle to opt-in crash shard on illegal state transitions

Default is logging an error and not crashing shard and all other entities, our tests have the toggle enabled.

Also: pulled in a fix for passivation when not using remember entities fixing #29359 and possibly #27549
as those were blocking this PR to test green

johanandren commented:
Added a toggle to opt-in crash shard on illegal state transitions
Default is logging an error and not crashing shard and all other entities, our tests have the toggle enabled.

Also: pulled in a fix for passivation when not using remember entities fixing #29359 and possibly #27549
as those were blocking this PR to test green

johanandren committed:
Test formatting...

johanandren committed:
mima filters

renatocaval commented on akka-cluster-sharding/src/multi-jvm/scala/akka/cluster/sbr/RandomizedBrainResolverIntegrationSpec.scala:
```diff
@@ -394,8 +394,11 @@ class RandomizedSplitBrainResolverIntegrationSpec
   val scenarios =
-    List(Scenario(leaseMajorityConfig, 3), Scenario(leaseMajorityConfig, 5), Scenario(leaseMajorityConfig, 9))
+    List(
+      Scenario(ConfigFactory.parseString("""test.random-seed = 1591321960290""").withFallback(leaseMajorityConfig), 3))
```
This look really random. How was this generated? 
https://xkcd.com/221/

johanandren replied:
Hehe, up, that was not intended for check in, was exploring another bug fail, good that you caught it

renatocaval commented on akka-cluster-sharding/src/test/scala/akka/cluster/sharding/EntitiesSpec.scala:
```diff
@@ -71,9 +76,10 @@ class EntitiesSpec extends AnyWordSpec with Matchers {
+      entities.rememberingStart("a", None) // need to go through remembering start to become active
+      entities.addEntity("a", ActorRef.noSender) // need to go through active to passivate
       entities.entityPassivating("a") // needs to go through passivating to be removed
+      entities.rememberingStop("a") // need to go through remembering start to become active
```
```suggestion
      entities.rememberingStop("a") // need to go through remembering stop to become active
```

octonato reviewed (approved):
LGTM

...with minor typo fix

johanandren committed:
Update akka-cluster-sharding/src/test/scala/akka/cluster/sharding/EntitiesSpec.scala

Co-authored-by: Renato Cavalcanti <renato@cavalcanti.be>

johanandren committed:
accidental commit reverted

johanandren commented:
Travis is green but not reported back, merging this.
